pandas/core/ops/array_ops.py

```diff
@@ -132,3 +152,171 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(left, right, op, str_rep, eval_kwargs):
```

Comment:
a docstring would be nice.

Reply:
Can you explain the flow a bit here? 
When do you end up here? What has already been done? 
Eg if you have a left and right where `left.__op__(right)` would return NotImplemeted, but `right.__op__(left)` works, do you end up here and in what form? 

Maybe a few examples of left and right types and how they are extracted before ending up here would help.

Reply:
> Can you explain the flow a bit here?
> When do you end up here? What has already been done?

This will all go in a docstring before long (hopefully in a separate pass, as this is a blocker for arithmetic-perf which I'm eager to get done).  The idea:

Everything `Series`-specific is done outside of this func (same with the comparison and logical funcs): potentially returning `NotImplemented`, finding res_name, checking/ensuring `Series` alignment (comparison is different from the other two), calling `_constructor` and `__finalize__`.

We should never get here (for any of these funcs) with a DataFrame, Series, or Index for `left`.  We should never get here with a DataFrame for right.  We can _almost_ rule out Series or Index for right, the unfortunate exception being L176-180.

Reply:
> This will all go in a docstring before long (hopefully in a separate pass, as this is a blocker for arithmetic-perf which I'm eager to get done)

Why not in this PR? Writing and reviewing a docstring should not take weeks, and it makes reviewing it here easier

Reply:
Doing a simple docstring would be easy, but since we are dealing with very central methods with a number of corner cases, doing a thorough docstring is not.  This PR is intended as a just-refactor, with a number of follow-ups when this blocker is done.